Bluetooth: Mesh: Shell: Split out mdl shell feat

Andrewpini opened the pull request:
This commit splits out the model related commands from mesh shell to
separate shell modules. The purpose of this alteration is to make model
commands more flexible by letting the application itself define the
model instances.

- Move Health Client commands to separate shell module.
- Move Config Client commands to separate shell module.
- Remove Health Client instance from shell.c.
- Add Health Client instance to mesh_shell test application.
- Add util library for Mesh shell.
- Some refactoring of shell menu layout.
- Update Mesh shell documentation.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV reviewed (commented):
Generally looks good. There is one thing I would like to propose. I would suggest to replace `SHELL_STATIC_SUBCMD_SET_CREATE` with `SHELL_SUBCMD_SET_CREATE` for `mesh_cmds`. With that change, it would let us move the health and client commands under `mesh` command. It would also allow to extended `mesh` command from any other source file. This would be beneficial for us in nRF Connect SDK as we would be able to move client's commands under `mesh` command. There is one downside for this. Commands consisted several words added directly to `mesh_cmds` will have to use underscore instead of dash. But this could be potentially be solved be converting them into nested commands.

You can see the proposal here: https://github.com/Andrewpini/zephyr/compare/align_mesh_shell...PavelVPV:zephyr:mesh_shell_subcmd_proposal?expand=1

Andrewpini commented:
> Generally looks good. There is one thing I would like to propose. I would suggest to replace `SHELL_STATIC_SUBCMD_SET_CREATE` with `SHELL_SUBCMD_SET_CREATE` for `mesh_cmds`. With that change, it would let us move the health and client commands under `mesh` command. It would also allow to extended `mesh` command from any other source file. This would be beneficial for us in nRF Connect SDK as we would be able to move client's commands under `mesh` command. There is one downside for this. Commands consisted several words added directly to `mesh_cmds` will have to use underscore instead of dash. But this could be potentially be solved be converting them into nested commands.
> 
> You can see the proposal here: https://github.com/Andrewpini/zephyr/compare/align_mesh_shell...PavelVPV:zephyr:mesh_shell_subcmd_proposal?expand=1

Yes, I think that is a good idea 👍

alxelax commented on subsys/bluetooth/mesh/shell_utils.c:
```diff
@@ -0,0 +1,102 @@
+{
+	if (argc == 1) {
+		shell_help(sh);
+		return 1;
```
Should it be any error from Zephyr's range?

Andrewpini replied:
I think it can return 0 instead

alxelax commented on tests/bluetooth/mesh_shell/src/main.c:
```diff
@@ -18,11 +18,44 @@ static struct bt_mesh_cfg_cli cfg_cli;
 
 BT_MESH_SHELL_HEALTH_PUB_DEFINE(health_pub);
 
+static void health_current_status(struct bt_mesh_health_cli *cli, uint16_t addr,
```
Since this is test I'd expect to see more detail information in shell that includes function incoming parameters for status messages.

Andrewpini commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -327,44 +326,80 @@ The Proxy Client model is an optional mesh subsystem that can be enabled through
+* Target the local node Configuration Server instance (``mdl_cfg target set <local_addr> <netidx>``).
+* Monitor the composition data of the local node (``mdl_cfg get-comp``).
+* Configure the local node as desired with the Configuration Client commands.
+* Provision other devices (``mesh beacon-listen``) (``mesh pb-adv``) (``mesh pb-gatt``).
```
Well, it could but the devices the shell cli module can provision is not restricted to only other devices with this shell module, so I thought it might be a little bit misleading.

alxelax replied:
Up to you. However, configuration module uses device key for app encryption\decryption. Only provisioner knows this. Naturally provisioner owns the config client. Otherwise we have to have some API to share device key.

Andrewpini replied:
I think I mention a bit above in the docs that the cfg client in most common cases will be dependent on the provisioner feature

Andrewpini commented on subsys/bluetooth/mesh/shell_config_cli.c:
```diff
@@ -0,0 +1,1712 @@
+	return 0;
+}
+
+int cmd_timeout(const struct shell *sh, size_t argc, char *argv[])
```
Probably overlooked from earlier. I will fix it 👍

PavelVPV reviewed (commented):
Consider that now there are 5 source files with `shell` prefix in mesh folder, how about creating a shell folder and moving files there? I would also suggest to rename `shell_config_cli.c` to `shell_cfg_cli.c` to align with `cfg_cli.c`.

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3335,227 +1215,146 @@ static int cmd_cdb_app_key_del(const struct shell *shell, size_t argc,
-}
+/* Placeholder for model shell modules that is configured in the application */
+SHELL_SUBCMD_SET_CREATE(model_cmds, (models));
+SHELL_SUBCMD_ADD((mesh), models, &model_cmds, "Model Commands", shell_model_cmds_help, 0,
```
I thought about it for a while. If the only supported models is the cfg cli and health cli the `models` subfolder would be a little bit unnecessary, but if we want to include our DS model support under the same shell instance I think  it would look neater if we had a subfolder for it. Otherwise the root cmd folder of `mesh` would potentially be cluttered again, which is what i wanted to avoid in the first place. It of course possible to keep the `models` subfolder anyhow, but the purpose of it would not be obvious for people exclusively using Zephyr.

Andrewpini replied:
> Otherwise, I guess `mesh_cmds` can be initialized with `SHELL_STATIC_SUBCMD_SET_CREATE`?

Is there a huge benefit of doing so? Retaining the possibility to add commands directly to `mesh` from other modules could come in handy.

PavelVPV replied:
> I thought about it for a while. If the only supported models is the cfg cli and health cli the models subfolder would be a little bit unnecessary, but if we want to include our DS model support under the same shell instance I think it would look neater if we had a subfolder for it. Otherwise the root cmd folder of mesh would potentially be cluttered again, which is what i wanted to avoid in the first place. It is of course possible to keep the models subfolder anyway, but the purpose of it would not be obvious for people exclusively using Zephyr.

Makes sense.

> Is there a huge benefit of doing so? Retaining the possibility to add commands directly to mesh from other modules could come in handy.

Yes, because commands added to `SHELL_STATIC_SUBCMD_SET_CREATE` can have dash in the name.

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3335,227 +1215,146 @@ static int cmd_cdb_app_key_del(const struct shell *shell, size_t argc,
-	/* Provisioning operations */
+SHELL_STATIC_SUBCMD_SET_CREATE(prov_cmds,
+	SHELL_CMD_ARG(uuid, NULL, "<UUID: 1-16 hex values>", cmd_uuid, 2, 0),
+	SHELL_CMD_ARG(input_num, NULL, "<Number>", cmd_input_num, 2, 0),
```
Im not sure if I understand. You meen that it could be: `SHELL_CMD_ARG(input num, NULL, "<Number>", cmd_input_num, 2, 0),`? Tried it for a command and it did not seem to work.

PavelVPV replied:
I mean the previous syntax "input-num" will work here. `SHELL_SUBCMD_ADD` doesn't allow to use dash, but `SHELL_CMD_ARG` does allow.

Andrewpini replied:
Thought it was the other way around 😅  Would you prefer dash?

PavelVPV replied:
Dash (or hyphen) is close to traditional style for long command names.

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3335,227 +1215,146 @@ static int cmd_cdb_app_key_del(const struct shell *shell, size_t argc,
-	SHELL_CMD_ARG(beacon-listen, NULL, "<val: off, on>", cmd_beacon_listen,
-		      2, 0),
+	/* Health Server Model Operations */
+	SHELL_CMD_ARG(health_srv_add_fault, NULL, "<Fault ID>", cmd_add_fault, 2, 0),
+	SHELL_CMD_ARG(health_srv_del_fault, NULL, "[Fault ID]", cmd_del_fault, 1, 1),
```
Unless you are familiar with the health models, `add_fault`  sounds quite ambiguous to me at least. The user could of course read the docs to get the purpose of the cmd, but I would prefer if it was possible to understand the purpose intuitively from the cmd menu as well.

PavelVPV replied:
The length of the command is a bit too long. Maybe add some nesting?

Andrewpini replied:
ok 👍

Andrewpini commented on subsys/bluetooth/mesh/shell_config_cli.c:
```diff
@@ -0,0 +1,1713 @@
+static struct {
+	uint16_t dst;
+	uint16_t net_idx;
+} net = {
```
Could you elaborate a bit?

PavelVPV replied:
Previously, this structure was used not only by config client model, but also by other commands (e.g. `net-send`, health client, etc.). It would be good to preserve that behavior.

Andrewpini replied:
Ok, I guess it could be preserved through utils?

PavelVPV replied:
If it fits there, why not?

alxelax commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -127,114 +130,122 @@ General configuration
+---------------------------------
 
-	Send a raw mesh message with the current destination address, network and application index. The message opcode must be encoded manually.
+	Enable or disable Low Power operation. Once enabled, the device will turn off its radio and start polling for friend nodes. The device will not be able to receive messages from the mesh network until the friendship has been established.
```
I'm pretty sure device can receive messages from the mesh network. It depends on configuration options. Even more, lpn device has to be configured after provisioning.

Andrewpini replied:
Sure 👍

alxelax commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -295,27 +306,34 @@ Provisioning
+Proxy
+=====
+
+The Proxy Server model is an optional mesh subsystem that can be enabled through the :kconfig:option:`CONFIG_BT_MESH_GATT_PROXY` configuration option.
```
```suggestion
The Proxy Server module is an optional mesh subsystem that can be enabled through the :kconfig:option:`CONFIG_BT_MESH_GATT_PROXY` configuration option.
```
`Model` is a strict BLE Mesh term. There is no Proxy model.

alxelax commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -324,47 +342,86 @@ The Proxy Client model is an optional mesh subsystem that can be enabled through
-``mesh friend [val: off, on]``
-------------------------------
+``mesh models cfg friend [val: off, on]``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Why did you change `----` by `^^^^`?

Andrewpini replied:
I added one additional level:  Models->Configuration Client. Before it was just Configuration Client. 
![image](https://user-images.githubusercontent.com/35562897/186902668-0c99b6f0-36bb-4acc-b237-e65108efbc06.png)

alxelax replied:
I didn't dig so deeply into doc rules. Thanks for explanation :)

alxelax reviewed (commented):
LGTM just minor comments to documentation

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -127,114 +130,122 @@ General configuration
 
-``mesh dst [destination address]``
-----------------------------------
+``mesh target dst [destination address]``
```
Should commands in documentation now be grouped the same way as in code? E.g. create a separate subsection for `mesh target` commands explaining why they are needed.

Andrewpini replied:
Hmm, I will take a look at it

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -127,114 +130,122 @@ General configuration
--------------------------------
+To allow a device to broadcast provisioning beacons, the :kconfig:option:`CONFIG_BT_MESH_PROV_DEVICE` configuration option must be enabled, along with either :kconfig:option:`CONFIG_BT_MESH_PB_GATT` or :kconfig:option:`CONFIG_BT_MESH_PB_ADV`.
+
+To allow a device to provision devices, either the :kconfig:option:`CONFIG_BT_MESH_PROVISIONER` or :kconfig:option:`CONFIG_BT_MESH_PB_GATT_CLIENT` configuration option must be enabled.
```
```suggestion
To allow a device to provision devices, the :kconfig:option:`CONFIG_BT_MESH_PROVISIONER` configuration option must be enabled.
```

`CONFIG_BT_MESH_PB_GATT_CLIENT` is needed for `mesh prov remote-gatt` command.

Andrewpini replied:
👍

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -127,114 +130,122 @@ General configuration
 
-``mesh pb-gatt <val: off, on>``
--------------------------------
+To allow a device to broadcast provisioning beacons, the :kconfig:option:`CONFIG_BT_MESH_PROV_DEVICE` configuration option must be enabled, along with either :kconfig:option:`CONFIG_BT_MESH_PB_GATT` or :kconfig:option:`CONFIG_BT_MESH_PB_ADV`.
```
Can this description be moved under corresponding commands?  `CONFIG_BT_MESH_PB_GATT` under `mesh prov pb-gatt` and `CONFIG_BT_MESH_PB_ADV` under `mesh prov pb-adv` command.

Generally it would be good if a description for each command would tell which Kconfig option is needed for the command to show up in the shell. Or will it overload the documentation? What do you think?

Andrewpini replied:
Yes, I think that is a good idea

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -324,47 +342,86 @@ The Proxy Client model is an optional mesh subsystem that can be enabled through
+* The shell module should now target itself.
+* Monitor the composition data of the local node (``mesh models cfg get-comp``).
+* Configure the local node as desired with the Configuration Client commands.
+* Provision other devices (``mesh prov beacon-listen``) (``mesh prov pb-adv``) (``mesh prov pb-gatt``).
```
I guess it should be:
```suggestion
* Provision other devices (``mesh prov beacon-listen``) (``mesh prov remote-adv``) (``mesh prov remote-gatt``).
```

Andrewpini replied:
👍

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -324,47 +342,86 @@ The Proxy Client model is an optional mesh subsystem that can be enabled through
+* Configure the node as desired with the Configuration Client commands.
+
+
+``mesh models cfg target set <Dst_addr> <NetKeyIndex>``
```
There is no such command anymore.

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -694,144 +751,138 @@ The Configuration Client uses the general messages parameters set by ``mesh dst`
 
-``mesh fault-get <Company ID>``
--------------------------------
+``mesh models health instance get_all``
```
```suggestion
``mesh models health instance get-all``
```

PavelVPV reviewed (changes requested):
Almost there. Please also rename `shell_config_cli.` to `shell_cfg_cli.c` (requested earlier). I'm not sure if `shell_` prefix is now needed for these files since they are in a separate folder.

Andrewpini commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -0,0 +1,1445 @@
+#endif
+	SHELL_CMD_ARG(rpl-clear, NULL, NULL, cmd_rpl_clear, 1, 0),
+
+	SHELL_CMD(health-srv, &health_srv_cmds, "Low Power commands", bt_mesh_shl_mdl_cmds_help),
```
I am inclined to not put them in models since these commands are dependent on the static shell implementation. I see that I have copy paste error on the description, so I will fix that. Do you not agree that these cmds are test-cmds? Any particular reason why you don't want them under test?

PavelVPV replied:
That is what confuses me. Though the implementation in shell.c of these commands look like a test commands, they are useless if you have you own health server. As well as when you have more than health server.

Andrewpini replied:
Right, I guess it will be tricky no matter where we put them. The only real solution that can be done for these commands is to clearly state in the documentation that you need to initialize the health server of the shell in the application to make this cmds work. an alternative would be a separate Kconfig option for this, so that we are certain that the user knowingly has initialized the health server and enabled the option. However, I'm not sure that this approach is the best.

PavelVPV replied:
I'll create a separate task to discuss how to fix this.

Andrewpini commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -108,7 +111,7 @@ The Bluetooth mesh shell commands are parsed with a variety of formats:
 ********
 
-The Bluetooth mesh shell implements a large set of commands. Some of the commands accept parameters, which are mentioned in brackets after the command name. For example, ``mesh lpn <value: off, on>``. Mandatory parameters are marked with angle brackets (e.g. ``<NetKeyIndex>``), and optional parameters are marked with square brackets (e.g. ``[destination address]``).
+The Bluetooth mesh shell implements a large set of commands. Some of the commands accept parameters, which are mentioned in brackets after the command name. For example, ``mesh low-power lpn <value: off, on>``. Mandatory parameters are marked with angle brackets (e.g. ``<NetKeyIndex>``), and optional parameters are marked with square brackets (e.g. ``[destination address]``).
```
👍

PavelVPV commented on include/zephyr/bluetooth/mesh/shell.h:
```diff
@@ -25,15 +25,22 @@ extern "C" {
 extern struct bt_mesh_health_srv bt_mesh_shell_health_srv;
 
-/** @brief External reference to health client */
-extern struct bt_mesh_health_cli bt_mesh_shell_health_cli;
```
I'm afraid we can't just remove this as `shell.h` is a public API (though not mentioned in Bluetooth Mesh Shell page). I think we should mark it as `__deprecated` and keep for 2 more releases (https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html?highlight=api+lifecycle#deprecated).

Andrewpini replied:
Dose this meen that the bt_mesh_shell_health_cli instance in shell.c must be preserved as an empty placeholder?

PavelVPV replied:
Actually, wait a minute, it seems like it can stay in shell.c (or shell_health_cli.c). The health cli shell commands will use the default instance or the one set by `instance set` command. `mesh_shell` can still use `bt_mesh_shell_health_cli` instance. But since `bt_mesh_shell_health_cli` was removed from the health cli shell commands then if `bt_mesh_shell_health_cli` is not used anywhere, it should be removed from the binary.

Andrewpini replied:
Not 100% sure if i got your point. Should we preserve it in shell.c (without any CB handlers)?

PavelVPV replied:
I think `bt_mesh_shell_health_cli` could stay in code without any changes. Unless we really want to get rid of it, but then it should be deprecated.

PavelVPV commented on include/zephyr/bluetooth/mesh/shell.h:
```diff
@@ -25,15 +25,30 @@ extern "C" {
 		BT_MESH_HEALTH_PUB_DEFINE(_name,                               \
 					  BT_MESH_SHELL_CUR_FAULTS_MAX);
 
+struct bt_mesh_shell_target {
```
Struct should also have description.

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -127,114 +129,130 @@ General configuration
+	* ``Fault ID``: ID of the fault to register (``0x0001`` to ``0xFFFF``)
+
+
+``mesh test health^srv del^fault [Fault ID]``
```
```suggestion
``mesh test health-srv del-fault [Fault ID]``
```

Andrewpini replied:
👍

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -127,114 +129,130 @@ General configuration
 
 	* ``val``: Enable or disable provisioning with advertiser
 
+To allow a device to provision devices, the :kconfig:option:`CONFIG_BT_MESH_PROVISIONER` configuration option must be enabled.
```
```suggestion
To allow a device to provision devices, the :kconfig:option:`CONFIG_BT_MESH_PROVISIONER` and :kconfig:option:`CONFIG_BT_MESH_PB_ADV` configuration options must be enabled.
```

Andrewpini replied:
👍

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -243,8 +261,10 @@ Provisioning
 
-``mesh provision-gatt <UUID> <NetKeyIndex> <addr> <AttentionDuration>``
------------------------------------------------------------------------
+To allow a device to provision devices over GATT, the :kconfig:option:`CONFIG_BT_MESH_PROVISIONER` and :kconfig:option:`CONFIG_BT_MESH_PB_GATT_CLIENT` configuration option must be enabled.
```
```suggestion
To allow a device to provision devices over GATT, the :kconfig:option:`CONFIG_BT_MESH_PROVISIONER` and :kconfig:option:`CONFIG_BT_MESH_PB_GATT_CLIENT` configuration options must be enabled.
```

Andrewpini replied:
👍

PavelVPV reviewed (commented):
> Please also rename `shell_config_cli.` to `shell_cfg_cli.c` (requested earlier). I'm not sure if `shell_` prefix is now needed for these files since they are in a separate folder.

This should be addressed.

Andrewpini committed:
Bluetooth: Mesh: Move shell.c to subfolder

Moves shell.c to shell subfolder. Done as a separate step to preserve
diff between previous and new revision of shell.c.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: Shell: Split out mdl shell feat

This commit splits out the model related commands from mesh shell to
separate shell modules. The purpose of this alteration is to make model
commands more flexible by letting the application itself define the
model instances.

- Move Health Client commands to separate shell module.
- Move Config Client commands to separate shell module.
- Remove Health Client instance from shell.c.
- Add Health Client instance to mesh_shell test application.
- Add util library for Mesh shell.
- Some refactoring of shell menu layout.
- Update Mesh shell documentation.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on tests/bluetooth/mesh_shell/src/main.c:
```diff
@@ -18,11 +18,64 @@ static struct bt_mesh_cfg_cli cfg_cli;
 
 BT_MESH_SHELL_HEALTH_PUB_DEFINE(health_pub);
 
+static void show_faults(uint8_t test_id, uint16_t cid, uint8_t *faults, size_t fault_count)
```
Could `bt_mesh_shell_health_cli` be still used here?

Andrewpini replied:
When building the test application I get deprecation warning when applying the old health cli instance. Should I remove deprecation tag in shell.h then?

PavelVPV replied:
Yes

PavelVPV commented:
Could you add a bit more explanation to subject line of the second commit, e.g.:
`Bluetooth: Mesh: Move shell.c` -> `Blietooth: Mesh: Move shell.c to shell/ subfolder`.
